File: ios/PiApp/PiApp/ViewModels/ChatViewModel.swift
import Foundation
import os
import SwiftUI
import PiAI
import PiAgentCore
import PiSession

private let logger = Logger(subsystem: "com.pi.ai", category: "ChatViewModel")

/// Represents a displayable message in the chat.
public struct ChatMessage: Identifiable {
    public let id: String
    public let role: String
    public let content: String
    public var thinking: String?
    public var toolCalls: [DisplayToolCall]
    public var isStreaming: Bool

    public init(id: String = UUID().uuidString, role: String, content: String, thinking: String? = nil, toolCalls: [DisplayToolCall] = [], isStreaming: Bool = false) {
        self.id = id
        self.role = role
        self.content = content
        self.thinking = thinking
        self.toolCalls = toolCalls
        self.isStreaming = isStreaming
    }
}

public struct DisplayToolCall: Identifiable {
    public let id: String
    public let name: String
    public var input: String
    public var output: String?
    public var isError: Bool
    public var details: ToolResultDetails?
    public var isComplete: Bool

    public init(id: String = UUID().uuidString, name: String, input: String = "", output: String? = nil, isError: Bool = false, details: ToolResultDetails? = nil, isComplete: Bool = false) {
        self.id = id
        self.name = name
        self.input = input
        self.output = output
        self.isError = isError
        self.details = details
        self.isComplete = isComplete
    }
}

@MainActor
@Observable
public final class ChatViewModel {
    public var messages: [ChatMessage] = []
    public var isStreaming = false
    public var streamingText = ""
    public var streamingThinking = ""
    public var activeToolCall: DisplayToolCall?
    public var currentModel: ModelDefinition?
    public var error: String?
    public var inputText = ""

    public var sessionId: String?
    public var session: Session?
    private var currentLeafId: String?

    private let agentRepository: AgentRepository

    public init(agentRepository: AgentRepository) {
        self.agentRepository = agentRepository
    }

    public func loadSession(_ session: Session) {
        self.session = session
        self.sessionId = session.id
        self.messages = []
        self.currentLeafId = nil

        // Load existing messages from the latest branch
        do {
            if let leaf = try agentRepository.getLatestLeaf(sessionId: session.id) {
                currentLeafId = leaf.id
                let entries = try agentRepository.getBranch(leafId: leaf.id)
                messages = entries.map { entry in
                    ChatMessage(
                        id: entry.id,
                        role: entry.role,
                        content: entry.content,
                        thinking: entry.thinking
                    )
                }
            }
        } catch {
            self.error = error.localizedDescription
        }

        // Set model from session
        let catalogue = agentRepository.getModelCatalogue()
        currentModel = catalogue.get(provider: session.modelProvider, id: session.modelId)
    }

    public func sendMessage() {
        let text = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, !isStreaming else { return }
        guard let model = currentModel, let sessionId = sessionId else {
            error = "No model or session selected"
            return
        }

        inputText = ""
        error = nil

        // Add user message
        let userMessage = ChatMessage(role: "user", content: text)
        messages.append(userMessage)

        // Start streaming
        isStreaming = true
        streamingText = ""
        streamingThinking = ""

        // Add placeholder for assistant response
        let assistantMessageId = UUID().uuidString
        messages.append(ChatMessage(id: assistantMessageId, role: "assistant", content: "", isStreaming: true))

        agentRepository.sendMessage(
            sessionId: sessionId,
            parentEntryId: currentLeafId,
            text: text,
            model: model,
            systemPrompt: session?.systemPrompt ?? "",
            onLeafUpdated: { [weak self] leafId in
                Task { @MainActor in
                    self?.currentLeafId = leafId
                }
            }
        ) { [weak self] event in
            guard let self = self else { return }
            Task { @MainActor in
                self.handleAgentEvent(event, assistantMessageId: assistantMessageId)
            }
        }
    }

    private func handleAgentEvent(_ event: AgentEvent, assistantMessageId: String) {
        guard let index = messages.firstIndex(where: { $0.id == assistantMessageId }) else { return }

        switch event {
        case .streamDelta(let delta):
            streamingText += delta
            messages[index] = ChatMessage(
                id: assistantMessageId,
                role: "assistant",
                content: streamingText,
                thinking: streamingThinking.isEmpty ? nil : streamingThinking,
                toolCalls: messages[index].toolCalls,
                isStreaming: true
            )

        case .thinkingDelta(let delta):
            streamingThinking += delta
            messages[index] = ChatMessage(
                id: assistantMessageId,
                role: "assistant",
                content: streamingText,
                thinking: streamingThinking,
                toolCalls: messages[index].toolCalls,
                isStreaming: true
            )

        case .toolCallStarted(let name, _):
            let toolCall = DisplayToolCall(name: name)
            activeToolCall = toolCall
            var updatedMsg = messages[index]
            updatedMsg.toolCalls.append(toolCall)
            messages[index] = updatedMsg

        case .toolCallCompleted(let name, let result):
            if let toolIndex = messages[index].toolCalls.lastIndex(where: { $0.name == name }) {
                messages[index].toolCalls[toolIndex].output = result.output
                messages[index].toolCalls[toolIndex].isError = result.isError
                messages[index].toolCalls[toolIndex].details = result.details
                messages[index].toolCalls[toolIndex].isComplete = true
            }
            activeToolCall = nil

        case .assistantMessage(let content, let thinking):
            streamingText = content
            if let thinking = thinking { streamingThinking = thinking }

        case .usageUpdate:
            break

        case .error(let message):
            error = message

        case .done:
            isStreaming = false
            messages[index] = ChatMessage(
                id: assistantMessageId,
                role: "assistant",
                content: streamingText,
                thinking: streamingThinking.isEmpty ? nil : streamingThinking,
                toolCalls: messages[index].toolCalls,
                isStreaming: false
            )
            streamingText = ""
            streamingThinking = ""
        }
    }

    public func cancelStreaming() {
        agentRepository.cancelStreaming()
        isStreaming = false
    }
}

File: android/pi-app/src/main/kotlin/com/pimobile/app/viewmodel/ChatViewModel.kt
package com.pimobile.app.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.pimobile.agent.AgentEvent
import com.pimobile.agent.AgentToolResult
import com.pimobile.ai.*
import com.pimobile.app.PiMobileApp
import com.pimobile.app.repository.AgentRepository
import com.pimobile.app.repository.ApiKeyRepository
import com.pimobile.tools.BuiltInTools
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull

data class ChatUiState(
    val sessionId: String? = null,
    val messages: List<ChatMessage> = emptyList(),
    val isStreaming: Boolean = false,
    val currentStreamText: String = "",
    val currentThinkingText: String = "",
    val activeToolCalls: List<ActiveToolCall> = emptyList(),
    val selectedModel: ModelDefinition? = null,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val error: String? = null
)

sealed class ChatMessage {
    data class User(val text: String) : ChatMessage()
    data class Assistant(val text: String, val thinking: String? = null) : ChatMessage()
    data class ToolCallMsg(val name: String, val input: JsonObject, val result: AgentToolResult? = null) : ChatMessage()
    data class ErrorMsg(val message: String) : ChatMessage()
}

data class ActiveToolCall(val name: String, val input: JsonObject)

class ChatViewModel(application: Application) : AndroidViewModel(application) {

    private val app = application as PiMobileApp

    private val apiKeyRepository = ApiKeyRepository(application)
    private val sandboxDir = application.getExternalFilesDir("sandbox")
        ?: application.filesDir.resolve("sandbox").also { it.mkdirs() }
    private val tools = BuiltInTools.create(application, sandboxDir)
    private val agentRepository = AgentRepository(
        agentLoop = app.agentLoop,
        sessionRepository = app.sessionRepository,
        extensionRegistry = app.extensionRegistry,
        apiKeyRepository = apiKeyRepository,
        tools = tools
    )

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    private var streamingJob: Job? = null
    private val conversationMessages = mutableListOf<Message>()

    fun initSession(sessionId: String) {
        // Restore last selected model from preferences
        val defaultProvider = apiKeyRepository.getDefaultProvider()
        val defaultModelId = apiKeyRepository.getDefaultModel()
        if (defaultProvider != null && defaultModelId != null) {
            val saved = app.modelCatalogue.get(defaultProvider, defaultModelId)
            if (saved != null) {
                _uiState.update { it.copy(selectedModel = saved) }
            }
        }

        if (sessionId == "new") {
            viewModelScope.launch {
                val session = app.sessionRepository.createSession(title = "New Chat")
                _uiState.update { it.copy(sessionId = session.id) }
            }
        } else {
            _uiState.update { it.copy(sessionId = sessionId) }
            loadExistingSession(sessionId)
        }
    }

    private fun loadExistingSession(sessionId: String) {
        viewModelScope.launch {
            val session = app.sessionRepository.getSession(sessionId) ?: return@launch
            val leafId = session.leafId
            if (leafId != null) {
                val entries = app.sessionRepository.getBranch(leafId)
                val messages = entries.mapNotNull { entry ->
                    when (entry.type) {
                        "message" -> {
                            val data = Json.parseToJsonElement(entry.data).jsonObject
                            val role = data["role"]?.jsonPrimitive?.contentOrNull
                            val content = data["content"]?.jsonPrimitive?.contentOrNull ?: ""
                            when (role) {
                                "user" -> ChatMessage.User(content)
                                "assistant" -> ChatMessage.Assistant(content)
                                else -> null
                            }
                        }
                        else -> null
                    }
                }
                _uiState.update { it.copy(messages = messages) }
            }
        }
    }

    fun setModel(model: ModelDefinition) {
        _uiState.update { it.copy(selectedModel = model) }
        apiKeyRepository.setDefaultProvider(model.provider)
        apiKeyRepository.setDefaultModel(model.id)
    }

    fun sendMessage(text: String) {
        val model = _uiState.value.selectedModel ?: return
        val sessionId = _uiState.value.sessionId ?: return

        _uiState.update { state ->
            state.copy(
                messages = state.messages + ChatMessage.User(text),
                isStreaming = true,
                currentStreamText = "",
                currentThinkingText = "",
                error = null
            )
        }

        streamingJob = viewModelScope.launch {
            try {
                val eventFlow = agentRepository.sendMessage(
                    sessionId = sessionId,
                    userMessage = text,
                    model = model,
                    systemPrompt = "You are Pi, a helpful AI assistant on a mobile device. You can use tools to help the user with tasks involving files, databases, web requests, and device media.",
                    existingMessages = conversationMessages.toList()
                )

                val textBuilder = StringBuilder()
                val thinkingBuilder = StringBuilder()

                eventFlow.collect { event ->
                    when (event) {
                        is AgentEvent.StreamDelta -> {
                            textBuilder.append(event.delta)
                            _uiState.update { it.copy(currentStreamText = textBuilder.toString()) }
                        }
                        is AgentEvent.ThinkingDelta -> {
                            thinkingBuilder.append(event.delta)
                            _uiState.update { it.copy(currentThinkingText = thinkingBuilder.toString()) }
                        }
                        is AgentEvent.AssistantMessage -> {
                            val msg = ChatMessage.Assistant(event.content, event.thinking)
                            _uiState.update { state ->
                                state.copy(
                                    messages = state.messages + msg,
                                    currentStreamText = "",
                                    currentThinkingText = ""
                                )
                            }
                            conversationMessages.add(Message(
                                role = Role.ASSISTANT,
                                content = MessageContent.Text(event.content),
                                thinking = event.thinking
                            ))
                            agentRepository.saveAssistantMessage(sessionId, event.content, event.thinking)
                            textBuilder.clear()
                            thinkingBuilder.clear()
                        }
                        is AgentEvent.ToolCallStarted -> {
                            _uiState.update { state ->
                                state.copy(
                                    messages = state.messages + ChatMessage.ToolCallMsg(event.name, event.input),
                                    activeToolCalls = state.activeToolCalls + ActiveToolCall(event.name, event.input)
                                )
                            }
                        }
                        is AgentEvent.ToolCallCompleted -> {
                            _uiState.update { state ->
                                val updatedMessages = state.messages.toMutableList()
                                val idx = updatedMessages.indexOfLast {
                                    it is ChatMessage.ToolCallMsg && it.name == event.name && it.result == null
                                }
                                if (idx >= 0) {
                                    val existing = updatedMessages[idx] as ChatMessage.ToolCallMsg
                                    updatedMessages[idx] = existing.copy(result = event.result)
                                }
                                state.copy(
                                    messages = updatedMessages,
                                    activeToolCalls = state.activeToolCalls.filter { it.name != event.name }
                                )
                            }
                            agentRepository.saveToolCall(
                                sessionId, event.name,
                                event.result.toolCallId,
                                event.result.output
                            )
                        }
                        is AgentEvent.UsageUpdate -> {
                            _uiState.update { state ->
                                state.copy(
                                    inputTokens = state.inputTokens + event.inputTokens,
                                    outputTokens = state.outputTokens + event.outputTokens
                                )
                            }
                        }
                        is AgentEvent.Error -> {
                            _uiState.update { state ->
                                state.copy(
                                    messages = state.messages + ChatMessage.ErrorMsg(event.message),
                                    error = event.message
                                )
                            }
                        }
                        is AgentEvent.Done -> {
                            _uiState.update { it.copy(isStreaming = false) }
                        }
                    }
                }

                conversationMessages.add(Message(
                    role = Role.USER,
                    content = MessageContent.Text(text)
                ))
            } catch (e: Exception) {
                _uiState.update { state ->
                    state.copy(
                        isStreaming = false,
                        error = e.message,
                        messages = state.messages + ChatMessage.ErrorMsg(e.message ?: "Unknown error")
                    )
                }
            }
        }
    }

    fun cancelStreaming() {
        streamingJob?.cancel()
        _uiState.update { it.copy(isStreaming = false) }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}
